android-components/components/feature/downloads/src/main/java/mozilla/components/feature/downloads/temporary/copy/CopyDownloadFeature.kt

```diff
@@ -0,0 +1,88 @@
+import mozilla.components.browser.state.action.ShareInternetResourceAction
+import mozilla.components.browser.state.selector.findTabOrCustomTabOrSelectedTab
+import mozilla.components.browser.state.state.content.ShareInternetResourceState
+import mozilla.components.browser.state.store.BrowserStore
+import mozilla.components.concept.fetch.Client
+import mozilla.components.feature.downloads.temporary.TemporaryDownloadFeature
+import mozilla.components.lib.state.ext.flowScoped
+import mozilla.components.support.ktx.android.content.copyImage
+import mozilla.components.support.ktx.kotlinx.coroutines.flow.ifChanged
+
+/**
+ * [Middleware] implementation for copying online resources.
+ *
+ * This will intercept only [ShareInternetResourceAction] [BrowserAction]s.
+ *
+ * Following which it will transparently
+ *  - download internet resources while respecting the private mode related to cookies handling
+ *  - temporarily cache the downloaded resources
+ *  - copy the resource to the device clipboard.
+ *
+ * with a 1 second timeout to ensure a smooth UX.
+ *
+ * To finish the process in this small timeframe the feature is recommended to be used only for images.
+ *
+ *  @param context Android context used for various platform interactions
+ *  @param store a reference to the application's [BrowserStore]
+ *  @param tabId ID of the tab session, or null if the selected session should be used.
+ *  @param rootView [View] of the root Fragment.
+ *  @param httpClient Client used for downloading internet resources
+ *  @param cleanupCacheCoroutineDispatcher Coroutine dispatcher used for the cleanup of old
+ *  cached files. Defaults to IO.
+ */
+class CopyDownloadFeature(
+    private val context: Context,
+    private val store: BrowserStore,
+    private val tabId: String?,
+    private val rootView: View,
+    httpClient: Client,
+    cleanupCacheCoroutineDispatcher: CoroutineDispatcher = Dispatchers.IO,
+) : TemporaryDownloadFeature(context, httpClient, cleanupCacheCoroutineDispatcher) {
+
+    override fun start() {
+        scope = store.flowScoped { flow ->
+            flow.mapNotNull { state -> state.findTabOrCustomTabOrSelectedTab(tabId) }
+                .ifChanged { it.content.copy }
+                .collect { state ->
+                    state.content.copy?.let { copyState ->
+                        logger.debug("Starting the copying process")
+                        startCopy(copyState)
+
+                        // This is a fire and forget action, not something that we want lingering the tab state.
+                        store.dispatch(CopyInternetResourceAction.ConsumeCopyAction(state.id))
+                    }
+                }
+        }
+    }
+
+    @VisibleForTesting
+    internal fun startCopy(internetResource: ShareInternetResourceState) {
+        val coroutineExceptionHandler = coroutineExceptionHandler("Copy")
+
+        scope?.launch(coroutineExceptionHandler) {
+            withTimeout(operationTimeoutMillis) {
```

Comment:
I see here using the same 1 second timeout.
This was needed for sharing because we wanted continuitym, it was important to show the share options for the image after at most 1 second of the user tapping "Share image".
In the case of copying an image to the clipboard I'd say that this timeout can be made bigger and would also think of this feature as a backup for when sharing could not complete in one second - users could still copy the image and until going to where they want to paste it we'd have more chances to download and add it to the clipboard.
Not really a blocker, wanted to just raise this as something to think about.

Reply:
Gotcha. Thanks for raising, that's a good point